src/opt/dom_info.cpp

```diff
@@ -0,0 +1,275 @@
+            instr2_dom_tree[p_ins] = iter1;
+        }
+    }
+}
+
+//  Compute dominator or post-dominator.  Instructions that do not use
+//  streams are left out.
+//
+void dom_info::compute_dom(bool reversed)
+{
+    std::size_t num_of_instrs = p_program->size();
+    if(num_of_instrs == 0)
+        return;
+    std::unordered_map<const instruction*, std::set<const instruction*>> instr2_doms;
+    std::unordered_map<const instruction*, int> instr2_points;
+    int cur_points   = reversed ? num_of_instrs - 1 : 0;
+    bool seen_stream = false;
+    program_visitor vis{p_program, reversed};
+    std::unordered_map<const instruction*, const instruction*>& instr2_dom_tree =
+        (reversed ? instr2_ipdom : instr2_idom);
+    for(auto ins = vis.begin(), end = vis.end();; ins = vis.next(ins))
+    {
+        const instruction* p_ins = &(*ins);
+        instr2_points[p_ins]     = cur_points;
+        if(ins->get_stream() < 0)
+        {
+            if(reversed)
+                cur_points--;
+            else
+                cur_points++;
+            ;
+            if(ins == end)
+                break;
+            continue;
+        }
+        seen_stream              = true;
+        const instruction* p_tmp = nullptr;
+        int cnt                  = 0;
+        // find dominators.
+        for(auto&& iter : vis.get_inputs(ins))
+        {
+            if(iter->get_stream() < 0)
+                continue;
+            const instruction* p_arg = &(*iter);
+            cnt++;
+            assert(instr2_doms.find(p_arg) != instr2_doms.end());
+            if(p_tmp == nullptr)
+                instr2_doms[p_ins] = instr2_doms[p_arg];
+            else
+                instr2_doms[p_ins] = set_intersection(instr2_doms[p_ins], instr2_doms[p_arg]);
+            p_tmp = p_arg;
+        }
+        // find immediate dominators.
+        if(cnt == 1)
+        {
+            instr2_dom_tree[p_ins] = p_tmp;
+        }
+        else if(cnt > 0)
+        {
+            std::unordered_map<const instruction*, const instruction*>& idom =
+                reversed ? instr2_ipdom : instr2_idom;
+            find_dom_tree(instr2_doms, p_ins, instr2_dom_tree, idom);
+        }
+
+        instr2_doms[p_ins].insert(p_ins);
+        if(ins == end)
+            break;
+        if(reversed)
+            cur_points--;
+        else
+            cur_points++;
+    }
+    if(seen_stream)
+    {
+        MIGRAPHX_DEBUG(dump_doms(instr2_points, reversed));
+    }
+}
+
+// Identify split points.  A split point has more than one
+// outputs that are executed in different streams.
+
+bool dom_info::is_split_point(instruction_ref ins)
+{
+    std::set<int> stream_set;
+    for(auto&& arg : ins->outputs())
+    {
+        int arg_stream = arg->get_stream();
+        if(arg_stream >= 0)
+            stream_set.insert(arg_stream);
+    }
+    if(stream_set.size() > 1)
+        return true;
+    return false;
+}
```

Comment:
This will be really slow, since it will always iterate every instruction even when its not necessary and it puts it in a temporary storage that is heap allocated. It would be better to just find if any of streams that are different from the first one:

```cpp
bool dom_info::is_split_point(instruction_ref ins)
{
    if (ins->outputs().empty())
        return false;
    return std::any_of(ins->outputs().begin()+1, ins->outputs().end(), [&](instruction_ref ins) {
        return ins->get_stream() != ins->outputs().front().get_stream();
    });
}
```

Reply:
My definition of split point is: there are more than one outputs that uses different streams.  
So if instruction 1 is assigned stream 0, it has one and only one output, and that output is stream 1.  Then it is not a split.   The original code was guided by has_mask.  Now "has_mask" is removed.

Reply:
> My definition of split point is: there are more than one outputs that uses different streams.

Yes, thats what the code above does. It checks if the outputs have different streams. 

> So if instruction 1 is assigned stream 0, it has one and only one output, and that output is stream 1. Then it is not a split.

Yes, the code above would return false in this case as it doesn't ever check the stream of the current instruction.

Reply:
An instruction's stream can be -1, means:  no stream is assigned.   You need to iterate outputs until a stream-assigned instruction is found, then compare the rest of stream-assigned outputs to this instruction.

Reply:
Also local object is allocated on stack, not on heap.

Reply:
> An instruction's stream can be -1, means: no stream is assigned. You need to iterate outputs until a stream-assigned instruction is found, then compare the rest of stream-assigned outputs to this instruction.

Ah yea, forgot that part. It could be something like this:

```cpp
bool dom_info::is_split_point(instruction_ref ins)
{
    int stream = -1;
    for(auto x:ins->outputs())
    {
        if (stream >= 0 and stream != x->get_stream())
            return true;
        stream = x->get_stream();
    }
    return false;
}
```

> Also local object is allocated on stack, not on heap.

Only the pointers are allocated on the stack. The default allocator(ie [`std::allocator`](https://en.cppreference.com/w/cpp/memory/allocator)) allocates memory on the heap.

Furthermore, `std::set` is a node-based which means it will allocate on every insertion.

Reply:
fixed